astropy/units/_typing.py

```diff
@@ -0,0 +1,195 @@
+# -*- coding: utf-8 -*-
```

Comment:
Should we just make this `typing`? We generally treat any module underneath the main submodule (`units` in this case) as semi-private, so from that perspective it seems fine. (A quick search shows no counter examples except deep somewhere in statistics, possibly just an oversight.)

Reply:
Also, could you add the default one-liner license comment?

Reply:
Done. I also added licensing notes about code copied from ``typing_extensions``. It's PSF, so we're good.

On a related note, pretty much all of this code could be deleted if ``typing_extensions`` were an astropy dependency. I've been looking into this and typing_extensions is

1. developed in the main repository of python.
2. developed by GVR and the core python developers.
3. compatible with any computer system that has ``pip``.

There a larger conversation to be had on how astropy wants to implement typing, but ``typing_extensions``` will  certainly be one of the decencies required to do this in a way compatible with python 3.7 and even 3.8. It doesn't seem so far fetched to make it a dependency now... 🤞